src/main/java/org/manifold/compiler/front/FunctionValueVertex.java

```diff
@@ -0,0 +1,158 @@
+package org.manifold.compiler.front;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.manifold.compiler.TypeTypeValue;
+import org.manifold.compiler.TypeValue;
+import org.manifold.compiler.Value;
+
+public class FunctionValueVertex extends ExpressionVertex {
+
+  private static Logger log = LogManager.getLogger(FunctionValueVertex.class);
+
+  private final ExpressionEdge functionTypeEdge;
+  public ExpressionEdge getFunctionTypeEdge() {
+    return this.functionTypeEdge;
+  }
+
+  private final ExpressionGraph functionBody;
+  public ExpressionGraph getFunctionBody() {
+    return this.functionBody;
+  }
+
+  public FunctionValueVertex(ExpressionGraph exprGraph,
+      ExpressionEdge functionTypeEdge, ExpressionGraph functionBody) {
+    super(exprGraph);
+    this.functionTypeEdge = functionTypeEdge;
+    this.functionTypeEdge.setTarget(this);
+    this.functionTypeEdge.setName("type");
+    this.functionBody = functionBody;
+  }
+
+  private FunctionTypeValue type = null;
+
+  @Override
+  public TypeValue getType() {
+    // this is a function, so return a function type
+    return this.type;
+  }
+
+  private FunctionValue function = null;
+  @Override
+  public Value getValue() {
+    return this.function;
+  }
+
+  @Override
+  public void elaborate() throws Exception {
+    if (function != null) {
+      return;
+    }
+    log.debug("elaborating function definition");
+    ExpressionVertex typeVertex = functionTypeEdge.getSource();
+    typeVertex.elaborate();
+    TypeValue typeConstructor = new FunctionTypeValue(
+        TypeTypeValue.getInstance(), TypeTypeValue.getInstance());
+    if (!(typeVertex.getValue() instanceof FunctionTypeValue)) {
+      throw new TypeMismatchException(
+          typeConstructor, typeVertex.getValue());
+    }
+    type = (FunctionTypeValue) typeVertex.getValue();
+    // inject input/output TupleValues
+    TupleTypeValue inputType = (TupleTypeValue) type.getInputType();
+    TupleTypeValue outputType = (TupleTypeValue) type.getOutputType();
+    Map<String, ExpressionEdge> inputEdges = new HashMap<>();
+    Map<String, ExpressionEdge> outputEdges = new HashMap<>();
+    for (String argName : inputType.getSubtypes().keySet()) {
+      ExpressionEdge e = new ExpressionEdge(null, null); // I know what I'm doing
+      functionBody.addEdge(e);
+      inputEdges.put(argName, e);
+    }
+    for (String argName : outputType.getSubtypes().keySet()) {
+      // "name resolution": look for a variable reference vertex with this name in the subgraph
+      for (Map.Entry<VariableIdentifier, VariableReferenceVertex> vRef
```

Comment:
This loop and the loop below might be cleaner if you build a `Map<String, VariableReferenceVertex> nameResolution` - this would be a lookup

What does this do if name resolution fails?